Introducción

En un centro meteorológico, se contratan sus servicios como Data Scientist para construir un modelo que prediga si lloverá o no en las próximas 24 horas, utilizando información de las 24 horas previas. Se dispone de un conjunto de 19 variables meteorológicas. La descripción de las variables es la siguiente:

MinTemp Temperatura mínima registrada.

MaxTemp Temperatura máxima registrada.

Lluvia Cantidad de lluvia registrada ese día en mm.

Evaporacion Evaporación (mm) en 24 horas.

Sol Número de horas de sol brillante en el día.

VelRafaga La velocidad (km/h) de la ráfaga de viento más fuerte en las 24 horas hasta la medianoche.

Vel9am La velocidad (km/h) de la ráfaga de viento a las 9am.

Vel3pm La velocidad (km/h) de la ráfaga de viento a las 9am.

Hum9am Porcentaje de humedad a las 9am.

Hum3pm Porcentaje de humedad a las 3pm.

Pres9am Presión atmosférica (hpa) a nivel del mar a las 9am.

Pre3pm Presión atmosférica (hpa) a nivel del mar a las 3pm.

Nub9am Fracción del cielo cubierto por nubes a las 9am. Se mide en “octavos”, de manera que un valor 0 indica cielo. totalmente despejado y 8, cielo totalmente cubierto.

Nub3pm Fracción del cielo cubierto por nubes a las 3pm. Se mide en “octavos”, de manera que un valor 0 indica cielo. totalmente despejado y 8, cielo totalmente cubierto.

Temp9am Temperatura en grados celsius a las 9am.

Temp3pm Temperatura en grados celsius a las 3pm.

LluviaHoy Variable indicadora que toma el valor 1 si la precipitación en mm. en las últimas 24 horas. excede 1 mm. y 0, si no.

Koppen Clasificación Koppen de la zona de medición (Temperate, Subtropical, Grassland, Tropical, Desert).

Estación Estación del Año.

LluviaMan Indicador de lluvia al día siguiente de la medición.

***Parte 1***

Cargue la base de datos Lluvia_full.csv, codifique la variable LluviaMan como 0 y 1 para los días sin lluvia y con lluvia respectivamente. 

Además, realice una separación de la base de datos en un set de entrenamiento y set de validación.
Utilice una proporción de 70:30 respectivamente estratificado por la LluviaMan. Para poder replicar sus resultados, fije una semilla antes de obtener los indices. 

Para ello utilice la función set.seed(2023).

In [4]:
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import style
import seaborn as sns
import numpy as np

np.set_printoptions(suppress=True)
pd.set_option('display.float_format', lambda x: f"{x:0.5f}")

plt.style.use("ggplot")
%config InlineBackend.figure_format = "retina"

In [5]:
lluvia = pd.read_csv('Lluvia_full (1) (1).csv')

In [6]:
lluvia.head()

MinTemp  MaxTemp  Lluvia  Evaporacion     Sol  VelRafaga   Vel9am   Vel3pm  \
0  7.60000 19.30000 0.00000      3.40000 9.40000   35.00000 13.00000 13.00000   
1  6.80000 19.50000 0.00000      3.40000 9.40000   46.00000 19.00000 28.00000   
2  5.50000 13.90000 0.00000      2.80000 3.80000   44.00000 26.00000 13.00000   
3  8.60000 14.30000 0.00000      2.80000 3.80000   35.00000 13.00000 17.00000   
4  7.90000 13.00000 0.00000      2.80000 3.80000   39.00000 15.00000 17.00000   

    Hum9am   Hum3pm    Pres9am     Pre3pm  Nub9am  Nub3pm  Temp9am  Temp3pm  \
0 73.00000 32.00000 1018.60000 1015.40000 1.00000 1.00000  9.40000 18.80000   
1 67.00000 42.00000 1018.40000 1015.20000 1.00000 1.00000 11.90000 17.90000   
2 72.00000 58.00000 1017.80000 1015.50000 7.00000 7.00000  8.10000 12.20000   
3 67.00000 56.00000 1017.40000 1015.00000 7.00000 7.00000  9.30000 13.40000   
4 68.00000 69.00000 1017.60000 1015.30000 7.00000 7.00000  9.00000 11.70000   

  LluviaHoy LluviaMan       Koppen  Estacion  
0        No        No  Subtropical  Invierno  
1        No        No  Subtropical  Invierno  
2        No        No    Temperate  Invierno  
3        No       Yes    Temperate  Invierno  
4        No        No    Temperate  Invierno

In [7]:
lluvia.isnull().sum()

MinTemp        0
MaxTemp        0
Lluvia         0
Evaporacion    0
Sol            0
VelRafaga      0
Vel9am         0
Vel3pm         0
Hum9am         0
Hum3pm         0
Pres9am        0
Pre3pm         0
Nub9am         0
Nub3pm         0
Temp9am        0
Temp3pm        0
LluviaHoy      0
LluviaMan      0
Koppen         0
Estacion       0
dtype: int64

In [8]:
#Procedemos a renombrar la columan LLuviaman con 0 y 1
lluvia['LluviaMan'] = lluvia['LluviaMan'].replace({'Yes':1,'No':0})

In [9]:
lluvia.head()


MinTemp  MaxTemp  Lluvia  Evaporacion     Sol  VelRafaga   Vel9am   Vel3pm  \
0  7.60000 19.30000 0.00000      3.40000 9.40000   35.00000 13.00000 13.00000   
1  6.80000 19.50000 0.00000      3.40000 9.40000   46.00000 19.00000 28.00000   
2  5.50000 13.90000 0.00000      2.80000 3.80000   44.00000 26.00000 13.00000   
3  8.60000 14.30000 0.00000      2.80000 3.80000   35.00000 13.00000 17.00000   
4  7.90000 13.00000 0.00000      2.80000 3.80000   39.00000 15.00000 17.00000   

    Hum9am   Hum3pm    Pres9am     Pre3pm  Nub9am  Nub3pm  Temp9am  Temp3pm  \
0 73.00000 32.00000 1018.60000 1015.40000 1.00000 1.00000  9.40000 18.80000   
1 67.00000 42.00000 1018.40000 1015.20000 1.00000 1.00000 11.90000 17.90000   
2 72.00000 58.00000 1017.80000 1015.50000 7.00000 7.00000  8.10000 12.20000   
3 67.00000 56.00000 1017.40000 1015.00000 7.00000 7.00000  9.30000 13.40000   
4 68.00000 69.00000 1017.60000 1015.30000 7.00000 7.00000  9.00000 11.70000   

  LluviaHoy  LluviaMan       Koppen  Estacion  
0        No          0  Subtropical  Invierno  
1        No          0  Subtropical  Invierno  
2        No          0    Temperate  Invierno  
3        No          1    Temperate  Invierno  
4        No          0    Temperate  Invierno

In [10]:
lluvia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56420 entries, 0 to 56419
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MinTemp      56420 non-null  float64
 1   MaxTemp      56420 non-null  float64
 2   Lluvia       56420 non-null  float64
 3   Evaporacion  56420 non-null  float64
 4   Sol          56420 non-null  float64
 5   VelRafaga    56420 non-null  float64
 6   Vel9am       56420 non-null  float64
 7   Vel3pm       56420 non-null  float64
 8   Hum9am       56420 non-null  float64
 9   Hum3pm       56420 non-null  float64
 10  Pres9am      56420 non-null  float64
 11  Pre3pm       56420 non-null  float64
 12  Nub9am       56420 non-null  float64
 13  Nub3pm       56420 non-null  float64
 14  Temp9am      56420 non-null  float64
 15  Temp3pm      56420 non-null  float64
 16  LluviaHoy    56420 non-null  object 
 17  LluviaMan    56420 non-null  int64  
 18  Koppen       56420 non-null  object 
 19  Esta

In [11]:
#Ahora procedemos a realizar la separacion de los datos
from sklearn.model_selection import train_test_split

data_y = lluvia['LluviaMan']
data_x = lluvia.drop(columns='LluviaMan')

x_train, x_test, y_train, y_test = train_test_split(data_x, data_y, train_size=0.7, random_state=2023, stratify=data_y)

In [12]:
x_train.head()

MinTemp  MaxTemp  Lluvia  Evaporacion      Sol  VelRafaga   Vel9am  \
26215 17.30000 27.00000 0.00000      5.20000 10.00000   59.00000  9.00000   
27325  4.90000 21.00000 0.00000      3.80000  6.30000   54.00000  7.00000   
20588 19.10000 27.60000 0.00000      9.40000  0.60000   63.00000 26.00000   
8470  11.70000 19.70000 0.00000      3.20000  9.50000   39.00000 20.00000   
56237  1.50000 20.90000 0.00000      2.40000  9.30000   20.00000  2.00000   

        Vel3pm   Hum9am   Hum3pm    Pres9am     Pre3pm  Nub9am  Nub3pm  \
26215 35.00000 63.00000 50.00000 1018.50000 1014.50000 1.00000 4.00000   
27325 31.00000 26.00000 76.00000 1016.30000 1014.90000 4.00000 6.00000   
20588 11.00000 75.00000 73.00000 1016.20000 1012.00000 7.00000 8.00000   
8470  22.00000 65.00000 44.00000 1024.70000 1021.10000 6.00000 7.00000   
56237  9.00000 80.00000 41.00000 1023.20000 1020.00000 1.00000 1.00000   

       Temp9am  Temp3pm LluviaHoy       Koppen   Estacion  
26215 22.90000 26.10000        No  Subtropical  Primavera  
27325 19.80000 14.60000        No    Temperate  Primavera  
20588 20.60000 23.20000        No  Subtropical  Primavera  
8470  13.20000 18.80000        No    Grassland   Invierno  
56237  8.40000 20.90000        No    Temperate      Otoño

In [13]:
x_test.head()

MinTemp  MaxTemp   Lluvia  Evaporacion      Sol  VelRafaga   Vel9am  \
37827 20.70000 35.30000  0.00000      9.80000 12.70000   54.00000 24.00000   
4725  16.50000 25.60000  1.20000      6.80000 10.60000   39.00000 20.00000   
30943 15.50000 22.50000 17.00000      6.80000  6.70000   39.00000 20.00000   
29315  6.50000 22.60000  0.00000      3.20000  9.40000   31.00000  9.00000   
35805  8.10000 14.60000  0.00000      3.40000  7.60000   50.00000 17.00000   

        Vel3pm   Hum9am   Hum3pm    Pres9am     Pre3pm  Nub9am  Nub3pm  \
37827 24.00000 56.00000 29.00000 1012.80000 1009.90000 1.00000 0.00000   
4725  20.00000 51.00000 42.00000 1017.10000 1016.20000 1.00000 2.00000   
30943 22.00000 67.00000 58.00000 1017.80000 1019.10000 3.00000 6.00000   
29315 15.00000 68.00000 39.00000 1031.00000 1027.40000 1.00000 1.00000   
35805 28.00000 75.00000 59.00000 1023.10000 1021.20000 1.00000 3.00000   

       Temp9am  Temp3pm LluviaHoy       Koppen  Estacion  
37827 25.20000 34.10000        No  Subtropical    Verano  
4725  19.90000 24.70000       Yes  Subtropical     Otoño  
30943 16.90000 21.20000       Yes    Temperate     Otoño  
29315 14.20000 21.80000        No  Subtropical     Otoño  
35805 10.90000 13.80000        No    Temperate  Invierno

In [14]:
variables_categoricas = x_train.columns[x_train.dtypes == "object"]
variables_numericas   = x_train.columns[x_train.dtypes != "object"]
print("Variables categóricas:", variables_categoricas)
print("Variables numéricas  :", variables_numericas)

Variables categóricas: Index(['LluviaHoy', 'Koppen', 'Estacion'], dtype='object')
Variables numéricas  : Index(['MinTemp', 'MaxTemp', 'Lluvia', 'Evaporacion', 'Sol', 'VelRafaga',
       'Vel9am', 'Vel3pm', 'Hum9am', 'Hum3pm', 'Pres9am', 'Pre3pm', 'Nub9am',
       'Nub3pm', 'Temp9am', 'Temp3pm'],
      dtype='object')


In [15]:
from sklearn.compose import make_column_transformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

# Primero instanciamos las clases
one_hot = OneHotEncoder()
scaler  = StandardScaler()
# Creamos transformer
#    (preprocesador, listado_de_variables)
transformer = make_column_transformer(
    (one_hot, variables_categoricas),          # Preprocesamiento para variables categóricas
    (scaler,  variables_numericas),            # Preprocesamiento para variables numéricas
    remainder                 = "passthrough", # Para mantener las variables que no se transforman
    verbose_feature_names_out = False          # Para no agregar el tipo de prep. a las variables
)
# Aprendemos sobre el set de entrenamiento
transformer.fit(x_train)

ColumnTransformer(remainder='passthrough',
                  transformers=[('onehotencoder', OneHotEncoder(),
                                 Index(['LluviaHoy', 'Koppen', 'Estacion'], dtype='object')),
                                ('standardscaler', StandardScaler(),
                                 Index(['MinTemp', 'MaxTemp', 'Lluvia', 'Evaporacion', 'Sol', 'VelRafaga',
       'Vel9am', 'Vel3pm', 'Hum9am', 'Hum3pm', 'Pres9am', 'Pre3pm', 'Nub9am',
       'Nub3pm', 'Temp9am', 'Temp3pm'],
      dtype='object'))],
                  verbose_feature_names_out=False)

In [16]:
transformer.get_feature_names_out()

array(['LluviaHoy_No', 'LluviaHoy_Yes', 'Koppen_Desert',
       'Koppen_Grassland', 'Koppen_Subtropical', 'Koppen_Temperate',
       'Estacion_Invierno', 'Estacion_Otoño', 'Estacion_Primavera',
       'Estacion_Verano', 'MinTemp', 'MaxTemp', 'Lluvia', 'Evaporacion',
       'Sol', 'VelRafaga', 'Vel9am', 'Vel3pm', 'Hum9am', 'Hum3pm',
       'Pres9am', 'Pre3pm', 'Nub9am', 'Nub3pm', 'Temp9am', 'Temp3pm'],
      dtype=object)

In [17]:
# Aplicamos cambios a set de entrenamiento y testeo
x_train_prep = pd.DataFrame(
    data    = transformer.transform(x_train),
    columns = transformer.get_feature_names_out()
)
x_test_prep = pd.DataFrame(
    data = transformer.transform(x_test),
    columns=transformer.get_feature_names_out()
)
# Resultados:
print("Set de entrenamiento preprocesado:")
display(x_train_prep.head())
print("Set de testeo preprocesado:")
display(x_test_prep.head())

Set de entrenamiento preprocesado:


LluviaHoy_No  LluviaHoy_Yes  Koppen_Desert  Koppen_Grassland  \
0       1.00000        0.00000        0.00000           0.00000   
1       1.00000        0.00000        0.00000           0.00000   
2       1.00000        0.00000        0.00000           0.00000   
3       1.00000        0.00000        0.00000           1.00000   
4       1.00000        0.00000        0.00000           0.00000   

   Koppen_Subtropical  Koppen_Temperate  Estacion_Invierno  Estacion_Otoño  \
0             1.00000           0.00000            0.00000         0.00000   
1             0.00000           1.00000            0.00000         0.00000   
2             1.00000           0.00000            0.00000         0.00000   
3             0.00000           0.00000            1.00000         0.00000   
4             0.00000           1.00000            0.00000         1.00000   

   Estacion_Primavera  Estacion_Verano  ...   Vel9am   Vel3pm   Hum9am  \
0             1.00000          0.00000  ... -0.80165  1.78728 -0.15407   
1             1.00000          0.00000  ... -1.04321  1.31749 -2.15451   
2             1.00000          0.00000  ...  1.25165 -1.03144  0.49473   
3             0.00000          0.00000  ...  0.52696  0.26047 -0.04593   
4             0.00000          0.00000  ... -1.64712 -1.26633  0.76506   

    Hum3pm  Pres9am   Pre3pm   Nub9am   Nub3pm  Temp9am  Temp3pm  
0  0.02121  0.18307 -0.04182 -1.15645 -0.11784  0.71328  0.49557  
1  1.30810 -0.13481  0.01632 -0.08391  0.63775  0.24255 -1.18545  
2  1.15961 -0.14926 -0.40524  0.98863  1.39335  0.36403  0.07166  
3 -0.27577  1.07893  0.91759  0.63112  1.01555 -0.75964 -0.57151  
4 -0.42425  0.86219  0.75769 -1.15645 -1.25124 -1.48850 -0.26454  

[5 rows x 26 columns]

Set de testeo preprocesado:


LluviaHoy_No  LluviaHoy_Yes  Koppen_Desert  Koppen_Grassland  \
0       1.00000        0.00000        0.00000           0.00000   
1       0.00000        1.00000        0.00000           0.00000   
2       0.00000        1.00000        0.00000           0.00000   
3       1.00000        0.00000        0.00000           0.00000   
4       1.00000        0.00000        0.00000           0.00000   

   Koppen_Subtropical  Koppen_Temperate  Estacion_Invierno  Estacion_Otoño  \
0             1.00000           0.00000            0.00000         0.00000   
1             1.00000           0.00000            0.00000         1.00000   
2             0.00000           1.00000            0.00000         1.00000   
3             1.00000           0.00000            0.00000         1.00000   
4             0.00000           1.00000            1.00000         0.00000   

   Estacion_Primavera  Estacion_Verano  ...   Vel9am   Vel3pm   Hum9am  \
0             0.00000          1.00000  ...  1.01008  0.49537 -0.53253   
1             0.00000          0.00000  ...  0.52696  0.02558 -0.80286   
2             0.00000          0.00000  ...  0.52696  0.26047  0.06220   
3             0.00000          0.00000  ... -0.80165 -0.56165  0.11626   
4             0.00000          0.00000  ...  0.16461  0.96515  0.49473   

    Hum3pm  Pres9am   Pre3pm   Nub9am   Nub3pm  Temp9am  Temp3pm  
0 -1.01820 -0.64054 -0.71051 -1.15645 -1.62904  1.06252  1.66498  
1 -0.37476 -0.01922  0.20530 -1.15645 -0.87344  0.25774  0.29093  
2  0.41718  0.08193  0.62686 -0.44142  0.63775 -0.19780 -0.22069  
3 -0.52325  1.98924  1.83340 -1.15645 -1.25124 -0.60779 -0.13298  
4  0.46667  0.84774  0.93213 -1.15645 -0.49564 -1.10888 -1.30239  

[5 rows x 26 columns]

In [18]:
y_test.value_counts()

0    13198
1     3728
Name: LluviaMan, dtype: int64

In [19]:
#procedemos a realizar la grilla de hiperparametros

#partimos con el SVC

from sklearn.model_selection import GridSearchCV, KFold
from sklearn.svm import SVC

# Definir los hiperparámetros a optimizar
hiperparametros = {"C": [0.1, 1, 10, 100], "kernel": ["rbf", "linear", "poly"]}

# Configurar la validación cruzada con KFold
configuraciones = KFold(n_splits=10, shuffle=True, random_state=2023)

# Crear el modelo SVM
svm = SVC(random_state=2023)

# Crear un objeto GridSearchCV
modelos = GridSearchCV(
    estimator=svm,              # Modelo a optimizar
    param_grid=hiperparametros,  # Hiperparámetros a buscar
    cv=configuraciones,          # Configuración de validación cruzada
    scoring="roc_auc",         # Métrica a optimizar (precisión)
    n_jobs=-1,                   # Utilizar todos los núcleos de CPU
    verbose=2                  # Mostrar información detallada
)

# Ajustar el modelo a los datos de entrenamiento
modelos.fit(x_train_prep, y_train)

# Mostrar los resultados de la búsqueda de hiperparámetros
print("Mejor configuración de hiperparámetros encontrada:")
print(modelos.best_params_)
print("Mejor puntuación (Roc Auc) en el conjunto de validación cruzada:")
print(modelos.best_score_)

Fitting 10 folds for each of 12 candidates, totalling 120 fits
[CV] END ..................................C=0.1, kernel=rbf; total time=  26.9s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.0s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.0s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.2s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.2s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.3s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.3s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.3s
[CV] END ..................................C=0.1, kernel=rbf; total time=  27.4s
[CV] END ..................................C=0.1, kernel=rbf; total time=  28.2s
[CV] END ...............................C=0.1, kernel=linear; total time=  26.4s
[CV] END ...............................C=0.1,

In [20]:
#ahora continuaremos con el modelo de abdol de clasificacion
from sklearn.tree import DecisionTreeClassifier

# Define los hiperparámetros que deseas optimizar
hiperparametros_arbol = {
    'min_samples_split': [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],  # Observaciones mínimas para división
    'ccp_alpha': [0.0, 0.01, 0.1],   # Costo de complejidad
    'max_depth': [None, 1, 2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]      # Profundidad máxima del árbol
}

# Crea un clasificador de árbol de clasificación
tree_classifier = DecisionTreeClassifier(random_state=2023)

# Crea un objeto GridSearchCV para buscar los mejores hiperparámetros
grid_search_arbol = GridSearchCV(
    estimator=tree_classifier,
    param_grid=hiperparametros_arbol,
    scoring='roc_auc',  # Métrica a optimizar (puedes cambiarla si lo deseas)
    cv=10,  # Número de divisiones para validación cruzada
    n_jobs=-1,  # Utilizar todos los núcleos de CPU disponibles
    verbose=2  # Mostrar información de progreso
)

# Ajusta el modelo en tus datos de entrenamiento
grid_search_arbol.fit(x_train_prep, y_train)

# Muestra los mejores hiperparámetros encontrados
print("Mejores hiperparámetros encontrados:")
print(grid_search_arbol.best_params_)
print("Mejor puntuación (Roc Auc) en el conjunto de validación cruzada:")
print(grid_search_arbol.best_score_)


Fitting 10 folds for each of 1260 candidates, totalling 12600 fits
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=1; total time=   0.0s
[CV] END .ccp_alpha=0.0, max_depth=None, min_samples_split=2; total time=   0.4s
[CV] END .ccp_alpha=0.0, max_depth=None, m

/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:425: FitFailedWarning: 
630 fits failed out of a total of 12600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
630 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 732, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python3.9/site-packages/sklearn/base.py", line 1144, in wrapper
    estimator._validate_params()
  File "/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python

Mejores hiperparámetros encontrados:
{'ccp_alpha': 0.0, 'max_depth': 7, 'min_samples_split': 17}
Mejor puntuación (Roc Auc) en el conjunto de validación cruzada:
0.86122184855518


In [21]:
from sklearn.ensemble import RandomForestClassifier


# Define los valores que deseas explorar para cada hiperparámetro
hiperparametros_randomForest = {
    'n_estimators': [ 350,400,450,500,550],  # Número de árboles
    'max_features': ['auto', 'sqrt', 'log2'],  # Número de predictores muestreados para cada árbol
    'min_samples_split': [2,3],  # Observaciones mínimas para división
}

# Crea el modelo Random Forest
rf = RandomForestClassifier(random_state=2023)

# Crea un objeto GridSearchCV
grid_search_randomForest = GridSearchCV(
    estimator=rf,            # Modelo a optimizar
    param_grid=hiperparametros_randomForest,    # Cuadrícula de hiperparámetros a buscar
    cv=10,                    # Número de divisiones para validación cruzada
    scoring='roc_auc',        # Métrica de evaluación (puedes cambiarla según tu necesidad)
    n_jobs=-1,                # Utilizar todos los núcleos de CPU
    verbose=2                 # Muestra información detallada
)

# Ajusta el modelo a los datos de entrenamiento
grid_search_randomForest.fit(x_train_prep, y_train)

# Muestra los resultados de la búsqueda de hiperparámetros
print("Mejor configuración de hiperparámetros encontrada:")
print(grid_search_randomForest.best_params_)
print("Mejor puntuación (ROC AUC) en el conjunto de validación cruzada:")
print(grid_search_randomForest.best_score_)

Fitting 10 folds for each of 30 candidates, totalling 300 fits
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=400; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=350; total time=   0.0s
[CV] END max_features=auto, min_samples_split=2, n_estimators=400; total time

/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:425: FitFailedWarning: 
100 fits failed out of a total of 300.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
20 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 732, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python3.9/site-packages/sklearn/base.py", line 1144, in wrapper
    estimator._validate_params()
  File "/Users/raimundoschulzehot/miniconda3/envs/diplomado/lib/python3.9

Mejor configuración de hiperparámetros encontrada:
{'max_features': 'log2', 'min_samples_split': 2, 'n_estimators': 550}
Mejor puntuación (ROC AUC) en el conjunto de validación cruzada:
0.8972521566851916


In [22]:
#procedemos a realizar los modelos para ada caso buscando el f1 score

#Ahora comenzamos con la Regrecion Logistica
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import  f1_score

regrecion_logistica= LogisticRegression(class_weight= "balanced").fit(X= x_train_prep, y= y_train)

regrecion_logistica_prob= regrecion_logistica.predict_proba(x_train_prep)[::,1]
regrecion_logistica_pred= regrecion_logistica.predict(x_test_prep)

print("f1_score Regrecion Logistica: ",f1_score(y_test,regrecion_logistica_pred)) 

f1_score Regrecion Logistica:  0.6478506296135476


In [23]:
#continuamos con el SVM

from sklearn.svm import SVC

svm = SVC(C=0.1, kernel="linear", degree =3, random_state=2023, probability = True, class_weight='balanced').fit(x_train_prep, y_train)

svm_prob= svm.predict_proba(x_test_prep)[::,1]
svm_pred= svm.predict(x_test_prep)


print("f1_score SVM: ",f1_score(y_test,svm_pred)) 


f1_score SVM:  0.6475923126754481


In [24]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score

# Crear un modelo de árbol de decisión
arbol_decision = DecisionTreeClassifier(max_depth=7, min_samples_split=17, random_state=2023, ccp_alpha=0.0, class_weight='balanced')

# Entrenar el modelo con los datos de entrenamiento procesados
arbol_decision.fit(x_train_prep, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred_arbol = arbol_decision.predict(x_test_prep)

# Calcular el valor F1
f1_arbol = f1_score(y_test, y_pred_arbol)

print(f'El valor F1 para el modelo de árbol de decisión es: {f1_arbol}')


El valor F1 para el modelo de árbol de decisión es: 0.6103190070776491


In [25]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score

# Crear un modelo de Random Forest
random_forest = RandomForestClassifier(n_estimators=550, max_features='log2', min_samples_split=2, random_state=2023, class_weight='balanced')

# Entrenar el modelo con los datos de entrenamiento procesados
random_forest.fit(x_train_prep, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred_rf = random_forest.predict(x_test_prep)

# Calcular el valor F1
f1_rf = f1_score(y_test, y_pred_rf)

print(f'El valor F1 para el modelo de Random Forest es: {f1_rf}')


El valor F1 para el modelo de Random Forest es: 0.6169657422512236


In [26]:

# Crear un diccionario con los valores F1
f1_scores = {
    'Modelo': ['Regresión Logística', 'SVM', 'Árbol de Decisión', 'Random Forest'],
    'F1 Score': [f1_score(y_test,regrecion_logistica_pred),f1_score(y_test,svm_pred), f1_arbol, f1_rf]
}

# Crear un DataFrame a partir del diccionario
f1_scores_df = pd.DataFrame(f1_scores)

# Mostrar la tabla
print(f1_scores_df)


                Modelo  F1 Score
0  Regresión Logística   0.64785
1                  SVM   0.64759
2    Árbol de Decisión   0.61032
3        Random Forest   0.61697


**Parte 3**

Compare los modelos ajustados utilizando el set de test, bajo el criterio de f1-score, ¿cuál es el mejor modelo en su caso?

El criterio del f1 score esta compuesto por dos elemntos, la precisión que es la proporción de datos correctamente clasificados como positivos entre todos los datos clasificados como positivos. En segundo lugar, el recall es la proporción de datos positivos correctamente clasificados entre todos los datos positivos. El F1 score se calcula como la media armónica de la precisión y el recall. 

Segun los f1 score obtenidos para los distintos modelos segun sus hiperparametros, definidos por nosotros, tenemos que:

**Modelo  F1 Score**

1)Regresión Logística   0.64785

2)SVM   0.64759

3)Árbol de Decisión   0.610323 

4)Random Forest   0.61697



En nuestro caso el modelo que mayor f1 score corresponde a la Regrecion Logistica, con un score de 0.64785, señalando que en este caso y metrica corresponde al mejor modelo, pues es el que tendria la mejor media armónica entre precisión y el recall. 
